tools/parse/src/main/scala/org/allenai/nlpstack/parse/poly/polyparser/ArcEagerTransitionSystem.scala

```diff
@@ -500,3 +281,86 @@ class ArcEagerGuidedCostFunction(
   }
 }
 
+/** The ArcEagerForbiddenArcInterpretation handles ForbiddenArc constraints for the arc-eager
+  * system. In other words, it translates these constraints into a function that returns true
+  * for any (state, transition) pair that violates the constraint.
+  *
+  * @param forbiddenArc the forbidden arc constraint to consider
+  */
+case class ArcEagerForbiddenArcInterpretation(
+    forbiddenArc: ForbiddenEdge
+) extends ParsingConstraintInterpretation {
+
+  def applyToParserState(state: TransitionParserState, transition: StateTransition): Boolean = {
+    state match {
+      case tpState: TransitionParserState =>
+        StackRef(0)(tpState).headOption match {
+          case Some(stackTop) => BufferRef(0)(tpState).headOption match {
+            case Some(bufferTop) => transition match {
+              case ArcEagerShift => false
+              case ArcEagerReduce => false
+              case _ => {
+                Set(forbiddenArc.token1, forbiddenArc.token2) == Set(stackTop, bufferTop)
+              }
+            }
+            case None => false
+          }
+          case None => false
+        }
+      case _ => false
```

Comment:
This can probably be written a little more monadic, with less nesting, but I'm not sure it's important.